qiita_db/meta_util.py

```diff
@@ -59,78 +58,97 @@ def _get_data_fpids(constructor, object_id):
-
-        # Next, the public artifacts
-        for artifact in qdb.artifact.Artifact.iter_public():
-            # Add the filepaths of the artifact
-            filepath_ids.update({fid for fid, _, _ in artifact.filepaths})
-
-            # Then add the filepaths of the prep templates
-            for pt in artifact.prep_templates:
-                filepath_ids.update({fid for fid, _ in pt.get_filepaths()})
-
-            # Then add the filepaths of the sample template
-            filepath_ids.update(
-                {fid
-                 for fid, _ in artifact.study.sample_template.get_filepaths()})
-
-        # Next, analyses
-        # Same as before, there are public, private, and shared
-        analyses = qdb.analysis.Analysis.get_by_status('public') | \
-            user.private_analyses | user.shared_analyses
-
-        for analysis in analyses:
-            filepath_ids.update(analysis.all_associated_filepath_ids)
-
-        return filepath_ids
+            return True
+
+        sql = """SELECT
+            (SELECT array_agg(artifact_id)
+             FROM qiita.artifact_filepath
+             WHERE filepath_id = {0}) AS artifact,
+            (SELECT array_agg(study_id)
+             FROM qiita.sample_template_filepath
+             WHERE filepath_id = {0}) AS sample_info,
+            (SELECT array_agg(prep_template_id)
+             FROM qiita.prep_template_filepath
+             WHERE filepath_id = {0}) AS prep_info,
+            (SELECT array_agg(job_id)
+             FROM qiita.job_results_filepath
+             WHERE filepath_id = {0}) AS job_results,
+            (SELECT array_agg(analysis_id)
+             FROM qiita.analysis_filepath
+             WHERE filepath_id = {0}) AS analysis""".format(filepath_id)
+        TRN.add(sql)
+
+        arid, sid, pid, jid, anid = TRN.execute_fetchflatten()
+
+        # artifacts
+        if arid:
+            # [0] cause we should only have 1
+            artifact = qdb.artifact.Artifact(arid[0])
+            if artifact.visibility == 'public':
+                return True
+            else:
+                # let's take the visibility via the Study
+                return artifact.study.has_access(user)
+        # sample info files
+        elif sid:
+            # the visibility of the sample info file is given by the
+            # study visibility
+            # [0] cause we should only have 1
+            return qdb.study.Study(sid[0]).has_access(user)
+        # prep info files
+        elif pid:
+            # the prep access is given by it's artifacts, if the user has
+            # access to any artifact, it should have access to the prep
+            # [0] cause we should only have 1
+            a = qdb.metadata_template.prep_template.PrepTemplate(
+                pid[0]).artifact
+            if (a.visibility == 'public' or a.study.has_access(user)):
+                return True
+            else:
+                for c in a.children:
```

Comment:
```python
for c in a.descendants.nodes():
```
`children` is only checking at the immediate children but not to the entire descendants.